openedx/core/djangoapps/xblock/api.py

```diff
@@ -0,0 +1,217 @@
+    """
+    # Is this block part of a course, a library, or what?
+    # Get the Learning Context Implementation based on the usage key
+    context_impl = get_learning_context_impl(usage_key)
+    # Now, using the LearningContextImpl and the Studio/LMS-specific logic,
+    # check if the block exists in this context and if the user has permission
+    # to render this XBlock view:
+    if parent_app.require_edit_permission():
+        authorized = context_impl.can_edit_block(user, usage_key)
+    else:
+        authorized = context_impl.can_view_block(user, usage_key)
+    if not authorized:
+        # We do not know if the block was not found or if the user doesn't have
+        # permission, but we want to return the same result in either case:
+        raise NotFound("XBlock {} does not exist, or you don't have permission to view it.".format(usage_key))
+
+    # TODO: load field overrides from the context
+    # e.g. a course might specify that all 'problem' XBlocks have 'max_attempts'
+    # set to 3.
+    # field_overrides = context_impl.get_field_overrides(usage_key)
+
+    runtime = get_runtime_system().get_runtime(user_id=user.id if user else None)
+
+    return runtime.get_block(usage_key)
+
+
+def get_block_metadata(block):
+    """
+    Get metadata about the specified XBlock
+    """
+    return {
+        "block_id": six.text_type(block.scope_ids.usage_id),
+        "block_type": block.scope_ids.block_type,
+        "display_name": get_block_display_name(block),
+    }
+
+
+def resolve_definition(block_or_key):
+    """
+    Given an XBlock, definition key, or usage key, return the definition key.
+    """
+    if isinstance(block_or_key, BundleDefinitionLocator):
+        return block_or_key
+    elif isinstance(block_or_key, BlockUsageKeyV2):
+        context_impl = get_learning_context_impl(block_or_key)
+        return context_impl.definition_for_usage(block_or_key)
+    elif isinstance(block_or_key, XBlock):
+        return block_or_key.scope_ids.def_id
+    else:
+        raise TypeError(block_or_key)
+
+
+def get_block_display_name(block_or_key):
+    """
+    Efficiently get the display name of the specified block. This is done in a
+    way that avoids having to load and parse the block's entire XML field data
+    using its parse_xml() method, which may be very expensive (e.g. the video
+    XBlock parse_xml leads to various slow edxval API calls in some cases).
+
+    This method also defines and implements various fallback mechanisms in case
+    the ID can't be loaded.
+
+    block_or_key can be an XBlock instance, a usage key or a definition key.
+
+    Returns the display name as a string
+    """
+    def_key = resolve_definition(block_or_key)
+    use_draft = parent_app.get_learning_context_params().get('use_draft')
+    cache = BundleCache(def_key.bundle_uuid, draft_name=use_draft)
+    cache_key = ('block_display_name', six.text_type(def_key))
+    display_name = cache.get(cache_key)
```

Comment:
Not for this PR, but I strongly suspect we're eventually going to want a multi-get version of this that can get things in batch.